README.md
# naptan: a python package

Get, view, and map NaPTAN data.

---

The package's functions return the data from the NaPTAN API as a pandas.DataFrame. This allows the data to be easily filtered, queried and manipulated. Once you have the data, you can export the data as a geoJSON, or view it via an interactive map.

```python
"""
Example code to view all the NaPTAN stops within Leicester and
Leicestershire on a map in the browser.
"""
import naptan

# get data as a dataframe
# 260 is Leicestershire and 269 is Leicester
leics_stops = naptan.get_area_stops(['260', '269'])

# View the stops in the browser on an interactive map.
naptan.view_map(leics_stops)
```

![coalville-example](docs/map-example.PNG)

## Installation

The latest release can be installed with:

```bash
$ pip install naptan
```

## Documentation

All documentation, tutorials, API reference etc can be found at the package's site: https://mullinscr.github.io/naptan/

## License

MIT. See LICENSE.txt.

## NaPTAN dataset

National public transport access nodes (NaPTAN) is a national dataset of all
public transport 'stops' in England, Scotland and Wales. Bus stops form the vast
majority of the dataset; but there is data on tram, metro, tube, rail, ferry and
air services too.

Included in the dataset, along with the unique identifiers, is common name,
locality, stop type, bearing of onward travel, and status (active, inactive or pending).

More information can be found on the UK government's website:

- [NaPTAN guide for data managers](https://www.gov.uk/government/publications/national-public-transport-access-node-schema/naptan-guide-for-data-managers)
- [NPTG and NaPTAN schema guide (2.5) (pdf)](http://naptan.dft.gov.uk/naptan/schema/2.5/doc/NaPTANSchemaGuide-2.5-v0.67.pdf)

docs/CHANGELOG.md
## [0.2.1](https://github.com/mullinscr/naptan/releases/tag/v0.2.1)

### Added

- Documentation and tutorials etc via this website (https://mullinscr.github.io/naptan/)


## [0.2.0](https://github.com/mullinscr/naptan/releases/tag/v0.2.0)

### Changed

- Re-write of the package to enable ease of working with the returned NaPTAN data (data now returned as a pandas.DataFrame).
-   `get_area_stops` now accepts a mix of numbers and strings

### Added

- `export_geojson`: export the stop data and its attributes as geoJSON file.
- `create_map`: create a folium.Map object displaying the stops and some key attributes.
- `view_map`: create the folium.Map as above, and display it in the browser.
- `naptan.ATCO_CODES`: a package level variable to display the ATCO codes to area lookup table

## [0.1.0](https://github.com/mullinscr/naptan/releases/tag/v0.1.0)

Initial commit. Didn't return stops within a dataframe, or provide the ability
to map them.

tests/test_naptan_api_requests.py
from importlib_resources import files
from unittest.mock import Mock

import pytest
import requests

import naptan
from naptan.naptan import APIError

def _read_response():
    good_response_stream = files('tests.data') \
        .joinpath('good-response.csv')
    with open(good_response_stream, 'rb') as f:
        return f.read()


class GoodResponseMock:
    status_code = 200
    reason = 'OK'
    content = _read_response()


class BadResponseMock:
    status_code = 400
    reason = 'Atco area codes or data format are incorrectly formatted'


@pytest.fixture
def good_response(monkeypatch):
    def mock_get(*args, **kwargs):
        return GoodResponseMock()
    monkeypatch.setattr(requests, "get", mock_get)

@pytest.fixture
def bad_response(monkeypatch):
    def mock_get(*args, **kwargs):
        return BadResponseMock()
    monkeypatch.setattr(requests, "get", mock_get)

@pytest.mark.parametrize('input_list', [
        ['2500DCL4060', '1100DEA10139', '068000000322'],
        ['250', '110', '68'],
        ('250', '110', '068'),
        ['250', '110', '68']
    ]
)
def test_format_stop_areas_with_stop_areas(input_list):
    assert naptan.naptan._format_stop_areas(input_list) == '068,110,250'

def test_format_stop_areas_set():
    areas = {'260', '080', '269'}
    sorted_areas = sorted(list(areas))
    assert naptan.naptan._format_stop_areas(sorted_areas) == '080,260,269'

def test_format_stop_areas_empty():
    stops = []
    assert naptan.naptan._format_stop_areas(stops) == ''

def test_format_stop_areas_non_string():
    areas = [260, 80, 269]
    with pytest.raises(TypeError):
        naptan.naptan._format_stop_areas(areas)

@pytest.mark.parametrize('inputs', [
        ['1100DEA10139', '1100DEA10138', '2500DCL4060'],
        ['110', '110', '250']]
)
def test_format_stop_areas_no_duplicates(inputs):
    assert naptan.naptan._format_stop_areas(inputs) == '110,250'

def test_format_stop_areas_type_mix():
    stops = [250010959, '1100EXT10']
    with pytest.raises(TypeError):
        naptan.naptan._format_stop_areas(stops)

def test_good_request(good_response):
    response = naptan.naptan._process_request('')
    assert response.status_code == 200

def test_bad_request(bad_response):
    with pytest.raises(APIError):
        naptan.naptan._process_request('abc')

def test_good_response():
    response = Mock(spec=requests.models.Response)
    response.content = _read_response()
    stop_df = naptan.naptan._process_response(response)
    assert stop_df.shape[0] == 15
    assert stop_df.iloc[0]['ATCOCode'] == '068000000754'
    assert stop_df.iloc[1]['Easting'] == 346553

def test_get_all_stops(good_response):
    stop_df = naptan.naptan.get_all_stops()
    assert stop_df.shape[0] == 15
    assert stop_df.iloc[0]['ATCOCode'] == '068000000754'
    assert stop_df.iloc[1]['Easting'] == 346553

def test_get_area_stops(good_response):
    stop_df = naptan.naptan.get_area_stops(['068', '110', '250'])
    assert stop_df.shape[0] == 15
    assert stop_df.iloc[7]['ATCOCode'] == '1100DEC11184'
    assert stop_df.iloc[8]['CommonName'] == "Bedland's Lane"

def test_get_specific_stops(good_response):
    stops = ['068000000754', '1100EXT10', '250020084']
    stop_df = naptan.naptan.get_specific_stops(stops)
    assert stop_df.shape[0] == 3
    assert stop_df.iloc[0]['ATCOCode'] == '068000000754'

def test_get_specific_stops_stops_not_present(good_response):
    stops = ['2500DCL4060', '1100DEA10139', '068000000322']
    stop_df = naptan.naptan.get_specific_stops(stops)
    assert stop_df.empty


tests/test_naptan_datetime_conversion.py
import datetime

import pytest

import naptan

@pytest.mark.parametrize('input', ['', 'abcd', '123', 123, 0.156])
def test_datetime_conversion_invalid_or_empty_string(input):
    assert naptan.naptan._str_to_datetime(input) is None

@pytest.mark.parametrize('input',
    [
    '0001-17-99T15:30:45',
    '2022-12-01T99:30:45',
    '2000-17-06T23:30:45',
    '1995-10-99T15:30:75',
    '2022-10-01T15:86:45'
    ]
)
def test_datetime_conversion_valid_format_but_invalid_date(input):
    assert naptan.naptan._str_to_datetime(input) is None

@pytest.mark.parametrize('input,valid_date',
    [
        ('2022-03-21T15:30:45', datetime.datetime(2022, 3, 21, 15, 30, 45, tzinfo=datetime.timezone.utc)),
        ('2022-03-21T15:30:45.123', datetime.datetime(2022, 3, 21, 15, 30, 45, 123, tzinfo=datetime.timezone.utc)),
        ('2022-03-21T15:30:45.123456', datetime.datetime(2022, 3, 21, 15, 30, 45, 123456, tzinfo=datetime.timezone.utc)),
        ('2022-03-21T15:30:45.123456789', datetime.datetime(2022, 3, 21, 15, 30, 45, 123456, tzinfo=datetime.timezone.utc)),
        ('2022-03-21T15:30:45+01:00', datetime.datetime(2022, 3, 21, 15, 30, 45, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
        ('2022-03-21T15:30:45-01:00', datetime.datetime(2022, 3, 21, 15, 30, 45, tzinfo=datetime.timezone(datetime.timedelta(seconds=-3600)))),
        ('2022-03-21T15:30:45+01:30', datetime.datetime(2022, 3, 21, 15, 30, 45, tzinfo=datetime.timezone(datetime.timedelta(seconds=+5400)))),
        ('2022-03-21T15:30:45.123456+01:00', datetime.datetime(2022, 3, 21, 15, 30, 45, 123456, tzinfo=datetime.timezone(datetime.timedelta(seconds=+3600)))),
        ('2022-03-21T15:30:45.123456+0100', datetime.datetime(2022, 3, 21, 15, 30, 45, 123456, tzinfo=datetime.timezone(datetime.timedelta(seconds=+3600))))
    ]
)
def test_datetime_conversion_valid_date(input, valid_date):
    assert naptan.naptan._str_to_datetime(input) == valid_date

naptan/naptan.py
from io import BytesIO
import json
import tempfile
from typing import Iterable, Optional, Union
import webbrowser

import folium
from folium.plugins.fast_marker_cluster import FastMarkerCluster
from folium.plugins import Search
import pandas as pd
from pyproj import CRS, Transformer
import requests


_BASE_URL = 'https://naptan.api.dft.gov.uk'
_API_VERSION = 'v1'


class APIError(Exception):
    """Raised for a status code != 200 returned from NaPTAN API"""


def _get_stops(area_codes: str = '', status: Optional[str] = None) -> pd.DataFrame:
    """Process the API request and returned response."""
    response = _process_request(area_codes)
    return _process_response(response, status)

def _process_request(area_codes: str) -> requests.models.Response:
    """
    Processes the API request.

    Args:
        area_codes: The formatted ATCO area codes string. Eg. '269,260,080'.

    Returns:
        The NaPTAN API response.

    Raises:
        APIError: If the response status code is not 200, then raise error with
            the status code and reason.
    """
    response = requests.get(f'{_BASE_URL}/{_API_VERSION}/access-nodes',
        timeout = 30,
        params = {
            'atcoAreaCodes': area_codes,
            'dataFormat': 'csv'
        }
    )

    if response.status_code != 200:
        raise APIError(response.status_code, response.reason)
    return response

def _convert_coords(df: pd.DataFrame) -> pd.DataFrame:
    """Ensures the DataFrame has a full compliment of Latitude and Longitude
    values. Fills this data in by converting the mandatory Easting and Northing
    values.

    Args:
        df: DataFrame of stops created by _process_response

    Returns:
        DataFrame with complete set of Latitude and Longitude values
    """
    crs_27700 = CRS('EPSG:27700')
    crs_4326 = CRS('EPSG:4326')
    transformer = Transformer.from_crs(crs_27700, crs_4326)
    df['Latitude'], df['Longitude'] = transformer.transform(df['Easting'], df['Northing'])
    return df

def _process_response(response: requests.models.Response, status: Optional[str]) -> pd.DataFrame:
    """
    Process the valid returned API response and return a dataframe.

    Args:
        response: The valid response from the NaPTAN API call.

        status: Return only NaPTAN stops with specific status. Must be one of:
            None, 'active', 'inactive', or 'pending'.

    Returns:
        All stops returned from valid response.
    """
    stop_df = pd.read_csv(
        BytesIO(response.content),
        dtype={c: str for c in [1, 2, 6, 21, 22, 23, 24, 25]}
    )

    stop_df = _convert_coords(stop_df)

    if status:
        status = status.lower().strip()
        if status not in ('active', 'inactive', 'pending'):
            raise ValueError("`status` must be one of: 'active', 'inactive', or 'pending'")
        return stop_df.query('Status == @status')
    return stop_df

def _format_stop_areas(stops: Iterable[Union[str, int]]) -> str:
    """
    Takes stops or stop areas and limits them to the first three characters.

    Will also handle area codes that have only been passed in as '89' or 89
    instead of '089', for example.

    Args:
        stops: Iterable of stops or area codes to be processed.

    Returns:
        Formatted area code string for passing to the API request, eg. '260,080'
    """
    stops = [str(stop) for stop in stops]
    areas = sorted(list({stop[:3].rjust(3, '0') for stop in stops}))
    return ','.join(areas)

def get_specific_stops(stops: Iterable[str], status: Optional[str] = None) -> pd.DataFrame:
    """
    Returns a dataframe containing just the specified stops, if present in the
    NaPTAN dataset.

    Args:
        stops: Iterable of the desired stops' ATCO code, as strings. For example:
            ['2500DCL4060', '1100DEA10139', '068000000322']

        status: Return only NaPTAN stops with specific status. Must be one of:
            None, 'active', 'inactive', or 'pending'. Defaults to None.

    Returns:
        The specified stops and their attributes.
    """
    stop_areas = _format_stop_areas(stops)
    returned_stops = _get_stops(stop_areas, status)
    return returned_stops.loc[returned_stops['ATCOCode'].isin(stops)]

def get_area_stops(area_codes: Iterable[Union[str, int]], status: Optional[str] = None) -> pd.DataFrame:
    """Returns a dataframe containing all the stops that share the specified area codes.

    Args:
        area_codes: Iterable of the desired area code. Can be strings or numeric.
            For example: ['250', '110'] or [250, 110]

        status: Return only NaPTAN stops with specific status. Must be one of:
            None, 'active', 'inactive', or 'pending'.

    Returns:
        All available stops and their attributes that share the specified area codes.
    """
    stop_areas = _format_stop_areas(area_codes)
    return _get_stops(stop_areas, status)

def get_all_stops(status: Optional[str] = None) -> pd.DataFrame:
    """Returns a dataframe with all the available nationwide NaPTAN stops.

    Args:
        status: Return only NaPTAN stops with specific status. Must be one of:
            None, 'active', 'inactive', or 'pending'. Defaults to None.

    Returns:
        All available NaPTAN stops and their attributes in a dataframe.
    """

    return _get_stops(status=status)

def _generate_geojson(df: pd.DataFrame) -> str:
    """Creates a geoJSON string from a stop dataframe.

    Args:
        df: Input data frame of stops returned from naptan functions.

    Returns:
        JSON (geoJSON) formatted string of the stops and their data.
    """
    stops = []

    for stop in df.itertuples():
        stop_geojson = {
                'type': 'Feature',
                'geometry': {
                    'type': 'Point',
                    'coordinates': [stop.Longitude, stop.Latitude],
                },
                'properties': {k: v for k, v in stop._asdict().items() if k != 'Index'}
        }
        stops.append(stop_geojson)

    geojson_str = {
        'type': 'FeatureCollection',
        'features': stops,
    }

    return json.dumps(geojson_str, indent=2)

def export_geojson(df: pd.DataFrame, path: str) -> None:
    """Export a dataframe of stops as a .json (geoJSON) file.

    Args:
        df: Input data frame of stops returned from naptan functions.
        path: Output path. e.g. '~/path/to/dir/naptan_stops.json'
    """
    geojson = _generate_geojson(df)
    with open(path, 'w') as f:
        f.write(geojson)

def create_map(df: pd.DataFrame, disable_cluster_zoom: int = 17) -> folium.Map:
    """Create a folium.Map object displaying the stops.

    Args:
        df: Input data frame of stops returned from naptan functions.
        disable_cluster_zoom: The zoom level at which point points are no longer
            clustered on the map, by default 17.

    Returns:
        A folium.Map object that can be used interactively or embedded in a webpage.
    """
    central_point = df['Latitude'].median(), df['Longitude'].median()
    m = folium.Map(location=[*central_point])
    data = [
        [lat, lng, atco, common_name, locality, ind, bearing]
        for lat, lng, atco, common_name, locality, ind, bearing in zip(
            df['Latitude'], df['Longitude'], df['ATCOCode'], df['CommonName'],
            df['LocalityName'], df['Indicator'], df['Bearing']
        )
    ]
    callback = """
        function (row) {
            var marker;
            marker = L.marker(
                new L.LatLng(row[0], row[1]),
                {"Identifier": row[2] + ": " + row[3] + ", " + row[4]}
            );
            marker.bindPopup(
                "<b>" + row[2] + "</b>"
                + "</br>"
                + row[4] + ", " + row[5] + " " + row[3]
                + "</br>"
                + "<a href='https://bustimes.org/stops/" + row[2] + "' target='_blank'>Bus Times (" + row[6] + ")</a>");
            return marker;
        };
    """
    mc = FastMarkerCluster(
        data,
        callback=callback,
        disableClusteringAtZoom=disable_cluster_zoom
    ).add_to(m)
    Search(layer=mc, search_label='Identifier', search_zoom=18, placeholder='Search by ATCO code...').add_to(m)
    return m

def view_map(df: pd.DataFrame, disable_cluster_zoom: int = 17) -> None:
    """View the stops on a folium-generated map, in the browser.

    Args:
        df: Input data frame of stops returned from naptan functions.
        disable_cluster_zoom: The zoom level at which point points are no longer
            clustered on the map, by default 17.
    """
    m = create_map(df, disable_cluster_zoom)
    f = tempfile.NamedTemporaryFile(mode='wb', delete=False) 
    m.save(f.name + '.html')
    webbrowser.open(f.name + '.html')